### ISSUE TITLE
Several same Forward References in TypedDict wrongly detected as Recusive Type

### ISSUE THREAD

**Issue by Jym77** [python/mypy#4812]
MNWE:
```
from mypy_extensions import TypedDict
from typing import Any, List

class Foo(TypedDict):
    bar: 'Bar'
    baz: 'Bar'

# Foo = TypedDict('Foo', {'bar': 'Bar', 'baz': Bar})

Bar = List[Any]
```

Result: 
```
$ mypy --python-version 3.6 mypy_test.py 
mypy_test.py:4: error: Recursive types not fully supported yet, nested types replaced with "Any"
```
(Python 3.6, mypy 0.570 & 0.580)

The TypedDict Foo has two fields, both being of type Bar which is defined later and a forward reference is used to mention it (obviously not needed in this MWNE but needed in real life). This is incorrectly detected as a Recursive type by mypy which then produce a rather confusing error message.

The error appears using both syntaxes for TypedDict (the class-like syntax used here, or the TypedDict() syntax commented out).

The error disappears without the forward reference (moving the definition of Bar up and removing the quotes in Foo). The error stays if Bar is moved up and the quote stay in Foo ("forward" reference backward?)

The error only appears if there are two (or more) fields of type Bar inside Foo. If either bar or baz is removed, the error disappears.

The error disappears if Bar is a simpler type, eg int or Any. It seems to depend on a more complex type.

**Comment by ilevkivskyi** [comment]
Good catch! I think there is a flaw in recursive type detection logic. Currently it fires if a type was encountered twice while processing a given node. The problem is that resolving subscripted forward type aliases results in the same type (even by identity). So it looks like we need a more sophisticated logic, or maybe a simpler solution would be to just copy the type every time we resolve a subscripted alias, this will be consistent with how we process simple (unsubscripted) aliases, that create a new instance for each encounter.

I am actually surprised this obvious corner case was not noticed before.

**Comment by Jym77** [comment]
Same problem (I guess), but even less intuitive:
```
from mypy_extensions import TypedDict
from typing import Any, List

Bar = List[Any]

class Foo(TypedDict):
    foo: 'Toto'
    bar: Bar
    baz: Bar

Toto = int
```
This causes an incorrect Recursive Types detection.
Removing any of the fields in Foo makes the Recursive Type detection disappear. Moving Toto up also does that.

So it needs both a type being used more than once, and a forward reference, but not necessarily on the same type. And the duplicated type need to be complex enough (`Bar = int` is OK).

**Comment by JukkaL** [comment]
This can probably also happen in import cycles -- added the label.

**Comment by cmekik** [comment]
I seem to have run into a similar issue, here is a minimal example:
```python
# python 3.7 w/ mypy 0.650 (as linter)
from typing import Union, Hashable, Any, Mapping, Set, NamedTuple

# In my actual code, MyBaseTuple is imported; but problem appears even w/o import
class MyBaseTuple(NamedTuple):

    base_field_1: Hashable
    base_field_2: Hashable


MyBaseTupleMapping = Mapping[MyBaseTuple, Any]
MyBaseTupleSet = Set[MyBaseTuple]

# Recursive types not fully supported yet, nested types replaced with "Any"
MyTupleUnion = Union['MyTupleA', 'MyTupleB']


class MyTupleA(NamedTuple):
    
    field_1: MyBaseTupleMapping
    field_2: MyBaseTuple


class MyTupleB(NamedTuple):
    
    field_1: MyBaseTupleMapping
    field_2: MyBaseTupleSet
    field_3: MyBaseTuple
```
Moving `MyTupleUnion` definition to after definition of `MyTupleA`, `MyTupleB` removes error.

**Comment by ilevkivskyi** [comment]
We will work on this beginning of the next year (we are going to rework the semantic analysis passes, so that import cycles and forward references will be fully supported).
